sopel/coretasks.py

```diff
@@ -543,101 +541,91 @@ def _parse_modes(bot, args, clear=False):
-            elif char in modes:
-                modes.pop(char)
-            param_idx += 1
-        elif char in chanmodes["C"]:
-            # Type C (l, etc) have a param only when setting
-            if sign == "+":
-                modes[char] = params[param_idx]
-                param_idx += 1
-            elif char in modes:
-                modes.pop(char)
-        elif char in chanmodes["D"]:
-            # Type D (aciLmMnOpqrRst, etc) have no params
-            if sign == "+":
-                modes[char] = True
-            elif char in modes:
-                modes.pop(char)
-        elif char in mapping and (
-            "PREFIX" not in bot.isupport or char in bot.isupport.PREFIX
-        ):
-            # User privs modes, always have a param
-            nick = Identifier(params[param_idx])
-            priv = channel.privileges.get(nick, 0)
-            # Log a warning if the two privilege-tracking data structures
-            # get out of sync. That should never happen.
-            # This is a good place to verify that bot.channels is doing
-            # what it's supposed to do before ultimately removing the old,
-            # deprecated bot.privileges structure completely.
-            ppriv = bot.privileges[channel_name].get(nick, 0)
-            if priv != ppriv:
-                LOGGER.warning(
-                    (
-                        "Privilege data error! Please share Sopel's "
-                        "raw log with the developers, if enabled. "
-                        "(Expected %s == %s for %r in %r)"
-                    ),
-                    priv,
-                    ppriv,
-                    nick,
-                    channel,
-                )
-            value = mapping.get(char)
-            if value is not None:
-                if sign == "+":
-                    priv = priv | value
-                else:
-                    priv = priv & ~value
-                bot.privileges[channel_name][nick] = priv
-                channel.privileges[nick] = priv
-            param_idx += 1
-        else:
-            # Might be in a mode block past A/B/C/D, but we don't speak those.
-            # Send a WHO to ensure no user priv modes we're skipping are lost.
+    privileges = mapping.keys()
+    if 'PREFIX' in bot.isupport:
+        privileges = bot.isupport.PREFIX.keys()
+
+    modemessage = ircmodes.ModeMessage(
+        bot.isupport.CHANMODES,
+        privileges=set(privileges),
+    )
+    modeinfo = modemessage.parse_modestring(args[1], tuple(args[2:]))
+
+    # set or update channel's modes
+    modes = {} if clear else copy.deepcopy(channel.modes)
+    for letter, mode, is_added, param in modeinfo.modes:
+        if letter == 'A':
+            if mode not in modes:
+                modes[mode] = set()
+            if is_added:
+                modes[mode].add(param)
+            elif param in modes[mode]:
+                modes[mode].remove(param)
```

Comment:
What about removing the mode when empty?

Just glancing at this for now. Full review to come after #2129 merges.

Reply:
Ha yes, the PR that will generate merge conflict on mine. 😁 

Yup, could be an option indeed, to remove the mode when empty.